Llava: add default chat templates

zucchini-nlp opened the pull request:
# What does this PR do?

This PR adds default chat templates for Llava models, so that user-defined models on the hub do not fail when using `apply_chat_template`. Users will see a warning in any case that the default template is used

Chat templates were added in all llava-hf models on the hub, can be verified by the below script

```python
messages = [{
            "role": "user",
            "content": [
                {"type": "text", "text": "What’s the content of this image?"},
                {"type": "image"},
                ],
        },
        {
            "role": "assistant",
            "content": [{"type": "text", "text": "This picture shows a red stop sign."},]
        }
    ]
processor = AutoProcessor.from_pretrained("llava-hf/vip-llava-7b-hf")

convo = processor.apply_chat_template(messages)
>>> ###Human: <image>\nWhat’s the content of this image?###Assistant: This picture shows a red stop sign.
```

zucchini-nlp committed:
add default chat templates

amyeroberts commented on src/transformers/models/llava/processing_llava.py:
````diff
@@ -134,3 +134,57 @@ def model_input_names(self):
+
+        Will create outputs like:
+        ```
+        USER: <image>\nWhat is the content of this image? ASSITANT: This picture shows a red stop sign
````
```suggestion
        USER: <image>\nWhat is the content of this image? ASSISTANT: This picture shows a red stop sign
```

amyeroberts commented on src/transformers/models/llava/processing_llava.py:
````diff
@@ -134,3 +134,57 @@ def model_input_names(self):
+
+        Will create outputs like:
+        ```
+        USER: <image>\nWhat is the content of this image? ASSITANT: This picture shows a red stop sign
````
Why is the image token being place before the text, when the text comes first in the content list? If this is the case, then this should be detailed above in bullet points of behaviour

zucchini-nlp replied:
Llava models need image to be placed before text, in all cases. I thought to make the template in a way that it format correctlly even if the users pass "image" in arbitrary order 

I can make it the same order as from user, then we need to update docs/readme on the hub explaining how to use the template correctly. WDYT, better to use user-order?

amyeroberts commented on src/transformers/models/llava/processing_llava.py:
```diff
@@ -134,3 +134,57 @@ def model_input_names(self):
+        This default vicuna template formats inputs in the form of a chat history. For each message in the chat history:
+        * the template will output the role of the speaker followed by the content of the message.
+        * content is a list of strings and images.
+        * If the content element is an image, the template will output a sequence of <image> tokens
```
Would be good to clarify what determines the number of image tokens here

amyeroberts commented on src/transformers/models/llava/processing_llava.py:
```diff
@@ -134,3 +134,57 @@ def model_input_names(self):
+        """
+        This default vicuna template formats inputs in the form of a chat history. For each message in the chat history:
+        * the template will output the role of the speaker followed by the content of the message.
+        * content is a list of strings and images.
```
This is ambiguous for two things: 
* Not clear if "content" refers to input i.e. "content" section in messages, or the output (as this comes after referring to the template outputting "content of the message". 
* Is it a list of strings and images? If input it's a list of dicts; if output it's strings

zucchini-nlp replied:
done, made more clear

amyeroberts commented on src/transformers/models/llava_next/processing_llava_next.py:
```diff
@@ -138,3 +138,57 @@ def model_input_names(self):
+    def default_chat_template(self):
+        """
+        This default vicuna template formats inputs in the form of a chat history. For each message in the chat history:
+        * the template will output the role of the speaker followed by the content of the message.
+        * content is a list of strings and images.
+        * If the content element is an image, the template will output a sequence of <image> tokens
```
Same comments here

amyeroberts commented on src/transformers/models/llava_next/processing_llava_next.py:
````diff
@@ -138,3 +138,57 @@ def model_input_names(self):
+
+        Will create outputs like:
+        ```
+        USER: <image>\nWhat is the content of this image? ASSITANT: This picture shows a red stop sign
````
```suggestion
        USER: <image>\nWhat is the content of this image? ASSISTANT: This picture shows a red stop sign
```

amyeroberts replied:
Same q here re image token being first

amyeroberts reviewed (commented):
Thanks for adding! 

Overall looks good, some nits, just needs some more clarifying language

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31691). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

zucchini-nlp committed:
Update src/transformers/models/llava/processing_llava.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update src/transformers/models/llava_next/processing_llava_next.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
more clear docstring and docs

NielsRogge commented on docs/source/en/model_doc/llava.md:
````diff
@@ -40,16 +40,38 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
-
-```bash
-"USER: <image>\n<prompt1> ASSISTANT: <answer1></s>USER: <prompt2> ASSISTANT: <answer2></s>USER: <prompt3> ASSISTANT:"
+- For better results, we recommend users to use processor's `apply_chat_template()` method to format your prompt correctly. Each message in the conversation history for chat templates is a dicts with keys "role" and "content". The "content" should be a list of dicts, for "text" and "image" modalaties, as follows:
````
```suggestion
- For better results, we recommend users to use the processor's `apply_chat_template()` method to format your prompt correctly. Each message in the conversation history for chat templates is a dictionary with keys "role" and "content". The "content" should be a list of dictionaries, for "text" and "image" modalities, as follows:
```

NielsRogge commented on docs/source/en/model_doc/llava_next.md:
```diff
@@ -46,25 +46,42 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
 - We advise users to use `padding_side="left"` when computing batched generation as it leads to more accurate results. Simply make sure to call `processor.tokenizer.padding_side = "left"` before generating.
 
-- Note that each checkpoint has been trained with a specific prompt format, depending on which large language model (LLM) was used. Below, we list the correct prompt formats to use for the text prompt "What is shown in this image?":
+- Note that each checkpoint has been trained with a specific prompt format, depending on which large language model (LLM) was used. You can use processor's `apply_chat_template` to format your prompts correctly. Each message in the conversation history for chat templates is a dicts with keys "role" and "content". The "content" should be a list of dicts, for "text" and "image" modalaties. Below is ...
```
```suggestion
- Note that each checkpoint has been trained with a specific prompt format, depending on which large language model (LLM) was used. You can use the processor's `apply_chat_template` to format your prompts correctly. Each message in the conversation history for chat templates is a dictionary with keys "role" and "content". The "content" should be a list of dictionaries, for "text" and "image" modalities. Below is an example of how to do that.
```

NielsRogge commented on docs/source/en/model_doc/vipllava.md:
````diff
@@ -34,16 +34,38 @@ Tips:
-
-```bash
-A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions.###Human: <image>\n<prompt1>###Assistant: <answer1>###Human: <prompt2>###Assistant:
+- For better results, we recommend users to use processor's `apply_chat_template()` method to format your prompt correctly. Each message in the conversation history for chat templates is a dicts with keys "role" and "content". The "content" should be a list of dicts, for "text" and "image" modalaties, as follows:
````
```suggestion
- For better results, we recommend users to use the processor's `apply_chat_template()` method to format your prompt correctly. Each message in the conversation history for chat templates is a dictionary with keys "role" and "content". The "content" should be a list of dictionaries, for "text" and "image" modalities, as follows:
```

NielsRogge reviewed (commented):
Would it be possible to add a `test_chat_template` test to https://github.com/huggingface/transformers/blob/main/tests/models/llava/test_processor_llava.py and so on?

Just an assertion which looks like this:

```python
prompt = "What is in this image?"

manually_formatted_prompt = f"USER: <image>\n{prompt} ASSISTANT:"

messages = ...

formatted_prompt = processor.apply_chat_template(messages, add_generation_prompt=True)

self.assertEquals(manually_formatted_prompt, formatted_prompt)
```

Cause I've seen that the templates are quite tricky to get 100% right at the space level.

zucchini-nlp committed:
Update docs/source/en/model_doc/llava.md

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/llava_next.md

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/vipllava.md

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

zucchini-nlp committed:
add tests

zucchini-nlp commented:
The tests will be failing until we add the templates on the hub. I will ping when we decide how to tackle the breaking change and it's is ready for review

zucchini-nlp committed:
remove default templates (see #31733)

zucchini-nlp committed:
load chat template from another file

zucchini-nlp committed:
Merge branch 'main' into chat_templates

zucchini-nlp commented:
Removed default chat templates (#31733 already removed all defaults) and added a hack to load chat templates from a separate json file called "chat_template.json". We can leave it as is for 2-3 major versions and then remove, because I guess everyone will have a version with `chat_template` in `init` by then. 

Ready for review @amyeroberts :)

amyeroberts reviewed (commented):
Thanks for working on adding this! 

We need to be careful about how we handle backwards and future compatibility with regards to loading the chat template. 

We'll also need to add general tests for the processor (perhaps in `test_processor_common.py`) which check the loading when a `chat_template.json` is or isn't present

zucchini-nlp committed:
Update docs/source/en/model_doc/llava_next.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp commented on docs/source/en/model_doc/llava_next.md:
````diff
@@ -46,25 +46,42 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
-```bash
-"A chat between a curious human and an artificial intelligence assistant. The assistant gives helpful, detailed, and polite answers to the human's questions. USER: <image>\nWhat is shown in this image? ASSISTANT:"
+# Note that the template simply formats your prompt, you still have to tokenize it and obtain pixel values for your images
+print(text_prompt)
````
oke, will bring that back

zucchini-nlp commented on src/transformers/processing_utils.py:
```diff
@@ -580,6 +584,23 @@ def get_processor_dict(
                     _raise_exceptions_for_missing_entries=False,
                 )
+
+                # load chat template from a separate json if exists
+                # TODO @raushan: remove this hack after a few major releases
```
Hmm I thought we could deprecate the `chat_template.json` when the old versions are very old to be in someone's env. I am okay with leaving `chat_template.json` but wouldn't that be different from what we have in LLMs? 

Another option I proposed was to add the template in processor's tokenizer. If we don't want to have the template in `processor.config`, this will be the better option IMO. WDYT?

zucchini-nlp replied:
> Older versions of transformers will still break as processors won't accept the chat template in the config.


Oh, also, continuing on this. We can also discuss this internally on slack later. We recently added an option for some processors to accept any kwargs and I was hoping to start integrating new kwargs for VLMs. Does this comment mean that we can't do it and will need another hack to load those kwargs?

amyeroberts replied:
> Hmm I thought we could deprecate the chat_template.json when the old versions are very old to be in someone's env

The problem is - when would this happen? There's some people who are still installing transformers v3, so we can't assume everyone will be working from the latest versions, even if it's a few months out. We can deprecate things within the code which we have full control over. Unfortunately, as the configs sit outside the library it means people can use them from the first version they're introduced until now. 

> Another option I proposed was to add the template in processor's tokenizer. If we don't want to have the template in processor.config, this will be the better option IMO. WDYT?

Just to be make sure we're talking about the same thing, this would mean adding into the tokenizer's config for the checkpoint e.g. [the tokenizer config here for a llava checkpoint](https://huggingface.co/llava-hf/llava-v1.6-mistral-7b-hf/tree/main)? 

It's a solution, but I don't think it semantically makes sense: the tokenizer is for processing text, whereas this introduces information about images.

> We recently added an option for some processors to accept any kwargs and I was hoping to start integrating new kwargs for VLMs. Does this comment mean that we can't do it and will need another hack to load those kwargs?

We definitely want this feature added for our processors, and gives us flexibility with future compatibility. It won't however fix things for older versions. My understanding is that if the chat_template gets added into the config, then older versions would still break when trying to load the config file.

zucchini-nlp replied:
> It's a solution, but I don't think it semantically makes sense: the tokenizer is for processing text, whereas this introduces information about images.

Yes, I meant that checkpoint. Now I see why it's not in tokenizer's config, then this workaround of having a separate `json` for templates is indeed better. 

>  It won't however fix things for older versions. My understanding is that if the chat_template gets added into the config, then older versions would still break when trying to load the config file.

Yeah, this will be the problem in any case as we won't have compatibility with older transformers versions given how breaking are the new changes. Dealing with hub-transformers compatibility is harder than it looked like 😅  I guess for that feature of VLM processor refactor we'll have to wait a while and try to change slowly, to see users reactions...

zucchini-nlp committed:
revert some changes in docs

zucchini-nlp committed:
forgot vipllava

zucchini-nlp committed:
Merge branch 'huggingface:main' into chat_templates

zucchini-nlp committed:
chat template file is not temporary hack

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into chat_templates

zucchini-nlp commented:
@amyeroberts I deleted a comment that it's a temporary hack, as we decoded to leave templates in their own json files. Ready for the final review

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into chat_templates

zucchini-nlp committed:
warn if loading from processor

zucchini-nlp commented:
I encountered something today on how we handle chat templates in some existing models (llava-next-video).

Llava-next-video was merged exactly at the moment when we merged chat-template PR and thus has it template in `processor_config`. Merging this PR will not break anything anywhere, but imo we need to be consistent on where the template is stored in the hub.

I'll upload a `chat_template.json` file on the hub for llava-next-video today. Now we're raising a warning if we find `processor_kwargs["chat_template"]` in any of the models, asking users to keep it in its own file.

zucchini-nlp committed:
not that file

zucchini-nlp committed:
similarly modify `save_pretrained`

amyeroberts commented on docs/source/en/model_doc/vipllava.md:
```diff
@@ -26,30 +26,64 @@ The abstract from the paper is the following:
+
+processor = AutoProcessor.from_pretrained("llava-hf/vip-llava-7b-hf")
+
+conversation = [
+    {
+        "role": "user",
+        "content": [
+            {"type": "image"},
+            {"type": "text", "text": "What’s shown in this image?"},
+            ],
+    },
+    {
+        "role": "assistant",
+        "content": [{"type": "text", "text": "This image shows a red stop sign."},]
+    },
+    {
+
+        "role": "user",
+        "content": [
+            {"type": "text", "text": "Decsribe the image in more details."},
+            ],
+    },
+]
```
nit - formatting here is a bit weird 

```suggestion
conversation = [
    {
        "role": "user",
        "content": [
            {"type": "image"},
            {"type": "text", "text": "What’s shown in this image?"},
         ,
    },
    {
        "role": "assistant",
        "content": [{"type": "text", "text": "This image shows a red stop sign."},]
    },
    {

        "role": "user",
        "content": [
            {"type": "text", "text": "Describe the image in more details."},
        ],
    },
]
```

amyeroberts commented on docs/source/en/model_doc/vipllava.md:
```diff
@@ -26,30 +26,64 @@ The abstract from the paper is the following:
+
+# Note that the template simply formats your prompt, you still have to tokenize it and obtain pixel values for your images
+print(text_prompt)
+>>> "###Human: <image>\nWhat’s shown in this image?###Assistant: This image shows a red stop sign.###Human: Decsribe the image in more details.###Assistant:"
```
```suggestion
>>> "###Human: <image>\nWhat’s shown in this image?###Assistant: This image shows a red stop sign.###Human: Describe the image in more details.###Assistant:"
```

amyeroberts commented on docs/source/en/model_doc/llava.md:
```diff
@@ -40,7 +40,42 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
+
+processor = AutoProcessor.from_pretrained("llava-hf/llava-1.5-7b-hf")
+
+conversation = [
+    {
+        "role": "user",
+        "content": [
+            {"type": "image"},
+            {"type": "text", "text": "What’s shown in this image?"},
+            ],
+    },
+    {
+        "role": "assistant",
+        "content": [{"type": "text", "text": "This image shows a red stop sign."},]
+    },
+    {
+
+        "role": "user",
+        "content": [
+            {"type": "text", "text": "Decsribe the image in more details."},
+            ],
+    },
+]
```
nit - same here re formatting 

```suggestion
conversation = [
    {
        "role": "user",
        "content": [
            {"type": "image"},
            {"type": "text", "text": "What’s shown in this image?"},
            ],
    },
    {
        "role": "assistant",
        "content": [{"type": "text", "text": "This image shows a red stop sign."},]
    },
    {

        "role": "user",
        "content": [
            {"type": "text", "text": "Describe the image in more details."},
        ],
    },
]
```

amyeroberts commented on docs/source/en/model_doc/llava.md:
```diff
@@ -40,7 +40,42 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
+
+# Note that the template simply formats your prompt, you still have to tokenize it and obtain pixel values for your images
+print(text_prompt)
+>>> "USER: <image>\n<What’s shown in this image? ASSISTANT: This image shows a red stop sign.</s>USER: Decsribe the image in more details. ASSISTANT:"
```
```suggestion
>>> "USER: <image>\n<What’s shown in this image? ASSISTANT: This image shows a red stop sign.</s>USER: Describe the image in more details. ASSISTANT:"
```

amyeroberts commented on docs/source/en/model_doc/llava_next.md:
```diff
@@ -46,26 +46,61 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
+
+        "role": "user",
+        "content": [
+            {"type": "text", "text": "Decsribe the image in more details."},
```
```suggestion
            {"type": "text", "text": "Describe the image in more details."},
```

amyeroberts commented on docs/source/en/model_doc/llava_next.md:
```diff
@@ -86,8 +121,17 @@ model.to("cuda:0")
+        "content": [
+            {"type": "image"},
+            {"type": "text", "text": "What is shown in this image?"},
+            ],
```
```suggestion
        ],
```

amyeroberts commented on docs/source/en/model_doc/llava_next.md:
```diff
@@ -46,26 +46,61 @@ The original code can be found [here](https://github.com/haotian-liu/LLaVA/tree/
+
+# Note that the template simply formats your prompt, you still have to tokenize it and obtain pixel values for your images
+print(text_prompt)
+>>> "[INST] <image>\nWhat's shown in this image? [/INST] This image shows a red stop sign. [INST] Decsribe the image in more details. [/INST]"
```
```suggestion
>>> "[INST] <image>\nWhat's shown in this image? [/INST] This image shows a red stop sign. [INST] Describe the image in more details. [/INST]"
```

amyeroberts commented on src/transformers/processing_utils.py:
```diff
@@ -494,11 +495,22 @@ def save_pretrained(self, save_directory, push_to_hub: bool = False, **kwargs):
                     del attribute.init_kwargs["auto_map"]
 
         # If we save using the predefined names, we can load using `from_pretrained`
+        # plus we save chat_template in itw own file
```
```suggestion
        # plus we save chat_template in its own file
```

amyeroberts commented on src/transformers/processing_utils.py:
```diff
@@ -494,11 +495,22 @@ def save_pretrained(self, save_directory, push_to_hub: bool = False, **kwargs):
+
+        processor_dict = self.to_dict()
+        if processor_dict.get("chat_template", None) is not None:
+            chat_template_dictionary = {"chat_template": processor_dict["chat_template"]}
```
I would have expected here that we pop `"chat_template"` so that when we save `processor_dict` out, the chat template won't be contained within the json file

zucchini-nlp replied:
oh yeah, we should pop. Bettter than the way I did now

amyeroberts commented on src/transformers/processing_utils.py:
```diff
@@ -557,14 +569,21 @@ def get_processor_dict(
+
         if os.path.isfile(pretrained_model_name_or_path):
             resolved_processor_file = pretrained_model_name_or_path
+            # cant't load chat-template when given a file as pretrained_model_name_or_path
+            resolved_chat_template_file = None
```
I'm not sure I understand this. If `pretrained_model_name_or_path` is a file, couldn't we check if there's a chat template file under the same path ?

zucchini-nlp replied:
My understanding was that if `pretrained_model_name_or_path` is a file, its a `processor_config.json` file. I might be wrong, as my usual way to load model is from the hub

amyeroberts replied:
ah, right, sorry, you're right

amyeroberts commented on src/transformers/processing_utils.py:
```diff
@@ -557,14 +569,21 @@ def get_processor_dict(
         elif is_remote_url(pretrained_model_name_or_path):
             processor_file = pretrained_model_name_or_path
             resolved_processor_file = download_url(pretrained_model_name_or_path)
+            # cant't load chat-template when given a file url as pretrained_model_name_or_path
```
```suggestion
            # can't load chat-template when given a file url as pretrained_model_name_or_path
```

amyeroberts commented on src/transformers/processing_utils.py:
```diff
@@ -594,6 +631,14 @@ def get_processor_dict(
+        chat_template = None
+        if resolved_chat_template_file is not None:
+            with open(resolved_chat_template_file, "r", encoding="utf-8") as reader:
+                text = reader.read()
+            chat_template = json.loads(text)["chat_template"]
```
Is there a reason for doing this in two steps i.e. getting `text` then `json.loads` instead of `json.load` in the reader context directly?

zucchini-nlp replied:
ah, will simplify it. I was copying from processor

zucchini-nlp replied:
Can't be loaded as `json` because it was saved as `TextIOWrapper`. Actually I don't know why we save it this way, was copying from processors. Maybe it has something to do with safe-saving 🤔

amyeroberts replied:
huh, interesting. Well, good to know :) Thanks for investigating

amyeroberts commented on src/transformers/processing_utils.py:
```diff
@@ -617,6 +662,12 @@ def get_processor_dict(
 
+        if "chat_template" in processor_dict and processor_dict["chat_template"] is not None:
+            logger.warning_once(
+                "Chat templates should be in a 'chat_template.json' file but found key='chat template' "
```
```suggestion
                "Chat templates should be in a 'chat_template.json' file but found key='chat_template' "
```

amyeroberts reviewed (approved):
Thanks for adding this! 

Pretty much all nits, apart from one question when `pretrained_model_name_or_path` is a file

zucchini-nlp committed:
Update tests/models/llava_next/test_processor_llava_next.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update tests/models/vipllava/test_processor_vipllava.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/vipllava.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update src/transformers/processing_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update src/transformers/processing_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/vipllava.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/llava.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/llava.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/llava_next.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/llava_next.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update src/transformers/processing_utils.py

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
Update docs/source/en/model_doc/llava_next.md

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

zucchini-nlp committed:
fix
